src/geode/inspector/criterion/intersections/surface_intersections.cpp

```diff
@@ -0,0 +1,404 @@
+/*
+ * Copyright (c) 2019 - 2023 Geode-solutions
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#include <geode/inspector/criterion/intersections/surface_intersections.h>
+
+#include <absl/algorithm/container.h>
+
+#include <geode/basic/logger.h>
+#include <geode/basic/pimpl_impl.h>
+
+#include <geode/geometry/aabb.h>
+#include <geode/geometry/basic_objects/segment.h>
+#include <geode/geometry/basic_objects/triangle.h>
+#include <geode/geometry/information.h>
+#include <geode/geometry/intersection_detection.h>
+
+#include <geode/mesh/core/triangulated_surface.h>
+#include <geode/mesh/helpers/aabb_surface_helpers.h>
+
+namespace
+{
+    geode::local_index_t vertex_position_to_index( geode::Position position )
+    {
+        if( position == geode::Position::vertex0 )
+        {
+            return 0;
+        }
+        if( position == geode::Position::vertex1 )
+        {
+            return 1;
+        }
+        if( position == geode::Position::vertex2 )
+        {
+            return 2;
+        }
+        return geode::NO_LID;
+    }
+
+    template < geode::index_t dimension >
+    class TriangleTriangleIntersection
+    {
+    public:
+        TriangleTriangleIntersection(
+            const geode::TriangulatedSurface< dimension >& mesh )
+            : mesh_( mesh )
+        {
+        }
+
+        void operator()( geode::index_t t1_id, geode::index_t t2_id )
+        {
+            if( t1_id == t2_id )
+            {
+                return;
+            }
+            for( const auto e : geode::LRange{ 3 } )
+            {
+                if( mesh_.polygon_adjacent( { t1_id, e } ) == t2_id )
+                {
+                    return;
+                }
+            }
+            const auto common_points = triangles_common_points( t1_id, t2_id );
+            if( common_points.size() == 3 )
+            {
+                return;
+            }
```

Comment:
two triangles having the same 3 vertices is not a problem? is it checked in another Inspector?